### PR TITLE
#145 - Move strings into Constants.swift

### PR DESCRIPTION
**Pull request by ywsang**
#145 
Added client-facing strings in Controllers to Constants.swift

### CODE REVIEW

File: TCAT/Supporting Files/Constants.swift [hunk 1]
```diff
@@ -54,22 +102,90 @@ struct Constants {
...
+
+        static let seeAllStops = "See All Stops"
+
+        // What's New phrases
+        static let whatsNewUpdateName = "App Shortcuts for Favorites"
+        static let whatsNewDescription = "Force Touch the app icon to search your favorites even faster."
     }
 
-    struct UserDefaults {
+    struct Stops {
```
Comment (mattbarker016) [inline TCAT/Supporting Files/Constants.swift]: Add this to Phrases, rename that to General

File: TCAT/Supporting Files/Constants.swift [hunk 2]
```diff
@@ -11,8 +11,51 @@ import UIKit
...
+        static let OK = "OK"
+
+        // Other button actions
+        static let retry = "Retry"
+        static let done = "Done"
+        static let magicSchoolBus = "✨📚🚌"
+        static let dismiss = "Dismiss"
+        static let begin = "BEGIN"
+        static let share = "Share"
+        static let back = "  Back"
```
Comment (mattbarker016) [inline TCAT/Supporting Files/Constants.swift]: Put space when this is called

File: TCAT/Supporting Files/Constants.swift [hunk 3]
```diff
@@ -27,13 +70,17 @@ struct Constants {
...
 
+    /// Banner titles
     struct Banner {
         static let noInternetConnection = "No internet connection"
         static let trackingLater = "Tracking available near departure time"
         static let cannotConnectLive = "Cannot connect to live tracking"
+        static let noLiveTrackingForRoutes = "No live tracking available for routes"
+        static let noLiveTrackingForRoute = "No live tracking available for Route "
+        static let cantConnectServer = "Could not connect to server"
+        static let calculationError = "Route calculation error. Please retry."
```
Comment (mattbarker016) [inline TCAT/Supporting Files/Constants.swift]: routeCalculationError

File: TCAT/Supporting Files/Constants.swift [hunk 4]
```diff
@@ -54,22 +102,90 @@ struct Constants {
...
+        static let searchAnywhere = "Search Anywhere."
+        static let favorites = "Your Favorites."
+        static let bestFeatures = "All the best features. All in one app."
+
+        // Detail label messages
+        static let welcomeMessage = "A beautiful and simple end-to-end navigation app for TCAT. Made by AppDev."
+        static let liveTrackingMessage = "Know exactly where your bus is and when it will be there."
+        static let searchAnywhereMessage = "From Ithaca Mall to Taughannock Falls, search any location and get there fast."
+        static let favoritesMessage = "All of your favorite destinations are just one tap away."
+        static let empty = ""
```
Comment (mattbarker016) [inline TCAT/Supporting Files/Constants.swift]: Remove this, but nice attention to detail

File: TCAT/Supporting Files/Constants.swift [hunk 5]
```diff
@@ -54,22 +102,90 @@ struct Constants {
...
+    }
+
     struct Phrases {
```
Comment (mattbarker016) [inline TCAT/Supporting Files/Constants.swift]: Rename to general

File: TCAT/Supporting Files/Constants.swift [hunk 6]
```diff
@@ -54,22 +102,90 @@ struct Constants {
...
     }
 
+    /// The phrases used in InformationViewController
+    struct InformationView {
+        static let onboarding = "Show Onboarding"
+        static let sendFeedback = "Send Feedback"
+        static let moreApps = "More Apps"
+        static let website = "Visit Our Website"
+        static let madeBy = "Made by Cornell App Development"
+        static let appDevDescription = "An Engineering Project Team\nat Cornell University"
```
Comment (mattbarker016) [inline TCAT/Supporting Files/Constants.swift:156]: Not ideal to be on this page. Any thoughts on how we could do this? Two labels maybe

**Review by mattbarker016 — COMMENTED** [review]
After you restructure the app around Alerts and finish the comments, LGTM. Would be good to get feedback from @myo3 and @AAAstorga just wait for the next commit

File: TCAT/Controllers/CustomNavigationController.swift [hunk 7]
```diff
@@ -9,171 +9,169 @@
...
         navigationBar.barTintColor = .white
         navigationBar.tintColor = .primaryTextColor
         navigationBar.titleTextAttributes = titleTextAttributes
         navigationItem.backBarButtonItem?.setTitleTextAttributes(
             CustomNavigationController.buttonTitleTextAttributes, for: .normal
         )
-        
+
         // Saved from other view controllers in case needed
         // navigationBar.isTranslucent = false
```
Comment (AAAstorga) [inline TCAT/Controllers/CustomNavigationController.swift:75]: can we delete this now?

File: TCAT/Controllers/CustomNavigationController.swift [hunk 8]
```diff
@@ -9,171 +9,169 @@
...
+            .font: UIFont.style(Fonts.SanFrancisco.regular, size: 14.0),
+            .foregroundColor: UIColor.primaryTextColor,
+            .baselineOffset: 0.3
         ]
-        let attributedString = NSMutableAttributedString(string: "  Back", attributes: attributes)
+        let attributedString = NSMutableAttributedString(string: Constants.Actions.back, attributes: attributes)
         backButton.setAttributedTitle(attributedString, for: .normal)
-        
+
         backButton.sizeToFit()
```
Comment (AAAstorga) [inline TCAT/Controllers/CustomNavigationController.swift:95]: don't have a space above this

File: TCAT/Controllers/CustomNavigationController.swift [hunk 9]
```diff
@@ -9,171 +9,169 @@
...
 
     }
-    
+
     override func popViewController(animated: Bool) -> UIViewController? {
-        
+
         let viewController = super.popViewController(animated: animated)
         if viewControllers.last is HomeViewController {
             viewControllers.last!.navigationItem.leftBarButtonItem = nil
```
Comment (AAAstorga) [inline TCAT/Controllers/CustomNavigationController.swift]: no force unwrap please

File: TCAT/Controllers/HomeViewController.swift [hunk 10]
```diff
@@ -535,7 +535,7 @@ extension HomeViewController: CLLocationManagerDelegate {
                 return
             }
 
-            let dontRemindAgainAction = UIAlertAction(title: "Don't Remind Me Again", style: .default) { (_) in
+            let dontRemindAgainAction = UIAlertAction(title: Constants.Actions.dontRemind, style: .default) { (_) in
```
Comment (AAAstorga) [inline TCAT/Controllers/HomeViewController.swift]:
```suggestion
            let dontRemindAgainAction = UIAlertAction(title: Constants.Actions.dontRemind, style: .default) { _ in
```

File: TCAT/Controllers/HomeViewController.swift [hunk 11]
```diff
@@ -578,7 +578,7 @@ extension HomeViewController: DZNEmptyDataSetSource, DZNEmptyDataSetDelegate {
         if isLoading {
```
Comment (AAAstorga) [inline TCAT/Controllers/HomeViewController.swift]: `if isLoading { return nil }`

File: TCAT/Controllers/SearchResultsTableViewController.swift [hunk 12]
```diff
@@ -297,16 +297,16 @@ extension SearchResultsTableViewController {
...
 
     func showLocationDeniedAlert() {
-        let alertController = UIAlertController(title: "Location Services Disabled",
-                                                message: "You need to enable Location Services in Settings",
+        let alertController = UIAlertController(title: Constants.AlertTitles.locationDisabled,
+                                                message: Constants.AlertMessages.enableLocation,
                                                 preferredStyle: .alert)
 
-        let settingsAction = UIAlertAction(title: "Settings", style: .default) { (_) in
+        let settingsAction = UIAlertAction(title: Constants.Actions.settings, style: .default) { (_) in
```
Comment (AAAstorga) [inline TCAT/Controllers/SearchResultsTableViewController.swift]:
```suggestion
        let settingsAction = UIAlertAction(title: Constants.Actions.settings, style: .default) { _ in
```

File: TCAT/Supporting Files/Constants.swift [hunk 13]
```diff
@@ -11,8 +11,51 @@ import UIKit
...
+        static let maxFavorites = "Maximum Number of Favorites"
+        static let couldntSendEmail = "Couldn't Send Email"
+        static let beep = "Be-beep be-beep!"
+        static let teleportation = "You're here!"
+        static let outOfRange = "Location Out Of Range"
+    }
 
+    struct App {
         /// The App Store Identifier used in App Store links
         static let storeIdentifier: String = "\(1290883721)"
```
Comment (AAAstorga) [inline TCAT/Supporting Files/Constants.swift]:
```suggestion
        static let storeIdentifier = "\(1290883721)"
```

**Review by AAAstorga — COMMENTED** [review]
Looks great to me. Looks like a big clean up. Good job :D

File: TCAT/Controllers/CustomNavigationController.swift (hunk 9 above)
Comment (ywsang) [reply TCAT/Controllers/CustomNavigationController.swift]: Tried to address this in the latest commit, but not sure if it was the best way to do it

File: TCAT/Controllers/CustomNavigationController.swift [hunk 14]
```diff
@@ -137,8 +136,10 @@ class CustomNavigationController: UINavigationController, UINavigationController
     override func popViewController(animated: Bool) -> UIViewController? {
 
         let viewController = super.popViewController(animated: animated)
-        if viewControllers.last is HomeViewController {
-            viewControllers.last!.navigationItem.leftBarButtonItem = nil
+        if let lastViewController = viewControllers.last {
```
Comment (mattbarker016) [inline TCAT/Controllers/CustomNavigationController.swift]: nitpick: Can combine into one line

`if let lastViewController = viewControllers.last as? HomeViewController { ... }`

**Comment by mattbarker016** [comment]
Really awesome job, Yana, especially with the organization tweaks! It just _feels_ cleaner now. After the nitpick, merge this on in! You might need to pull master first, I pushed that one quick fix for the placeholder empty pin that wasn't supposed to be there.

**Pull request merged**